File: Recursion/powerset.js
function powerset(s) {
  // base case
  if (s.length == 0) return [[]];
  // recursive step
  const without_first = powerset(s.slice(1));
  // subsets with first elem
  const with_first = without_first.map((el) => [s[0]].concat(el));
  // return
  return with_first.concat(without_first);
}
console.log(powerset(["A", "B", "C", "D"]));

File: Recursion/powerset.py
def power_set(my_list):
    # base case: an empty list
    if len(my_list) == 0:
        return [[]]
    # recursive step: subsets without first element
    without_first = power_set(my_list[1:])
    # subsets with first element
    with_first = [ [my_list[0]] + rest for rest in without_first ]
    # return combination of the two
    return with_first + without_first
  
universities = ['A', 'B', 'C', 'D']
power_set_of_universities = power_set(universities)
print(power_set_of_universities)